persistent-postgresql/test/JSONTest.hs

```diff
@@ -40,431 +42,654 @@ insert' :: (MonadIO m, PersistStoreWrite backend, BaseBackend backend ~ SqlBacke
+  objEmptyK <- insert' $ object ["" .= Number 9001]
+  objFullK  <- insert' $ object ["a" .= Number 1, "b" .= Number 2
+                                , "c" .= Number 3, "d" .= Number 4 ]
+  return $ EdgeCaseKeys {..}
+
+globalKeys :: (Monad m, MonadIO m) => ReaderT SqlBackend m (GlobalKeys TestValue)
+globalKeys = do
+          nullK <- insert' Null
+          let nulls = [nullK]
+
+          boolTK <- insert' $ Bool True
+          boolFK <- insert' $ toJSON False
+          let bools = [boolTK, boolFK]
+
+          num0K <- insert' $ Number 0
+          num1K <- insert' $ Number 1
+          numBigK <- insert' $ toJSON (1234567890 :: Int)
+          numFloatK <- insert' $ Number 0.0
+          numSmallK <- insert' $ Number 0.0000000000000000123
+          numFloat2K <- insert' $ Number 1.5
+          -- numBigFloatK will turn into 9876543210.123457 because JSON
+          numBigFloatK <- insert' $ toJSON (9876543210.123456789 :: Double)
+          let numbers = [ num0K, num1K, numBigK, numFloatK
+                        , numSmallK, numFloat2K, numBigFloatK ]
+
+          strNullK <- insert' $ String ""
+          strObjK <- insert' $ String "{}"
+          strArrK <- insert' $ String "[]"
+          strAK <- insert' $ String "a"
+          strTestK <- insert' $ toJSON ("testing" :: Text)
+          str2K <- insert' $ String "2"
+          strFloatK <- insert' $ String "0.45876"
+          let strings = [ strNullK, strObjK, strArrK, strAK
+                        , strTestK, str2K, strFloatK ]
+
+          arrNullK <- insert' $ Array $ V.fromList []
+          arrListK <- insert' $ toJSON [emptyArr,emptyArr,toJSON [emptyArr,emptyArr]]
+          arrList2K <- insert' $ toJSON [emptyArr,toJSON [Number 3,Bool False]
+                                        ,toJSON [emptyArr,toJSON [Object mempty]]
+                                        ]
+          arrFilledK <- insert' $ toJSON [Null, Number 4, String "b"
+                                         ,Object mempty, emptyArr
+                                         ,object [ "test" .= [Null], "test2" .= String "yes"]]
+          let lists = [ arrNullK, arrListK, arrList2K, arrFilledK ]
+
+          objNullK <- insert' $ Object mempty
+          objTestK <- insert' $ object ["test" .= Null, "test1" .= String "no"]
+          objDeepK <- insert' $ object ["c" .= Number 24.986, "foo" .= object ["deep1" .= Bool True]]
+          let objects = [ objNullK, objTestK, objDeepK ]
+          return GlobalKeys{..}
+
+
+data EdgeCaseKeys record =
+  EdgeCaseKeys { arrList3K  :: Key record
+               , arrList4K  :: Key record
+               , objEmptyK  :: Key record
+               , objFullK   :: Key record }
+
+data GlobalKeys record =
+  GlobalKeys { nulls   :: [Key record]
+             , bools   :: [Key record]
+             , numbers :: [Key record]
+             , strings :: [Key record]
+             , lists   :: [Key record]
+             , objects :: [Key record] }
```

Comment:
Formatting wise, It hink this should be:

```haskell
data GlobalKeys record = GlobalKeys
  { nulls :: [Key record]
  , bools :: [Key record]
  , ...
  }
```

I'm curious why it's polymorphic.